### ISSUE TITLE
Inconsistent behaviour of ImmutableMatrix

### ISSUE THREAD

**Issue by oscarbenjamin** [symengine/symengine.py#363]
The SymEngine tests in SymPy CI were not running properly. I'm reenabling them in https://github.com/sympy/sympy/pull/21992 but have found some possible regressions in SymEngine to do with the behaviour of ImmutableMatrix.

This is as expected:
```python
$ USE_SYMENGINE=1 python
Python 3.9.0 (v3.9.0:9cf6752276, Oct  5 2020, 11:29:23) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.backend import Matrix, ImmutableMatrix, sin, cos
>>> M = Matrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> M
[-1 + sin(1)**2 + cos(1)**2]

>>> print(M.simplify())
None
>>> M
[0]
```
However this is not expected:
```python
>>> IM = ImmutableMatrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> IM
[-1 + sin(1)**2 + cos(1)**2]

>>> print(IM.simplify())
None
>>> IM
[0]
```
It should not be possible to mutate the immutable matrix. SymPy's ImmutableMatrix class returns a simplified copy and does not mutate self:
```python
>>> from sympy import ImmutableMatrix
>>> SIM = ImmutableMatrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> SIM
Matrix([[-1 + cos(1)**2 + sin(1)**2]])
>>> print(SIM.simplify())
Matrix([[0]])
>>> SIM
Matrix([[-1 + cos(1)**2 + sin(1)**2]])
```
It would have been better if the mutable matrix method `simplify` had a different name because it is inconsistent with almost all the other `.simplify()` methods in SymPy. For compatibility though the SymEngine wrapper should match this behaviour.

This would have shown up in SymPy CI if the SymEngine tests were still running properly (I noticed this from the old Travis CI).

Another inconsistency is in binary operations that mix and match mutable and immutable matrices. With SymPy we have:
```python
>>> from sympy import Matrix, ImmutableMatrix
>>> M1 = Matrix([[1]])
>>> M2 = ImmutableMatrix([[1]])
>>> for M in [M1, M2, M1+M2, M2+M1]: print(type(M))
... 
<class 'sympy.matrices.dense.MutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
```
In other words all operations mixing mutable and immutable coerce to immutable. With SymEngine that instead gives:
```python
>>> from sympy.core.backend import Matrix, ImmutableMatrix
>>> M1 = Matrix([[1]])
>>> M2 = ImmutableMatrix([[1]])
>>> for M in [M1, M2, M1+M2, M2+M1]: print(type(M))
... 
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.ImmutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.ImmutableDenseMatrix'>
```
So the type of `M1+M2` does not match `M2+M1`.

**Comment by oscarbenjamin** [comment]
Also in SymPy a mutable matrix sympifies to an immutable matrix. Both kinds of SymEngine matrices sympify to *mutable* SymPy matrices. They should instead sympify to immutable SymPy matrices:
```python
In [1]: from sympy import Matrix, sympify

In [2]: type(sympify(Matrix([[1]])))
Out[2]: sympy.matrices.immutable.ImmutableDenseMatrix

In [3]: from sympy.core.backend import Matrix, ImmutableMatrix

In [4]: type(sympify(Matrix([[1]])))
Out[4]: sympy.matrices.dense.MutableDenseMatrix

In [5]: type(sympify(ImmutableMatrix([[1]])))
Out[5]: sympy.matrices.dense.MutableDenseMatrix
```

**Comment by isuruf** [comment]
Agree with all of the above. PRs are welcome.

**Comment by moorepants** [comment]
We started seeing numerous new symengine errors in this PR https://github.com/sympy/sympy/pull/24977. It seemed to me that at least some of the errors were not necessarily related to the changes made in that PR (removind Vector.simp = True), but I'm not sure. We also then started seeing new errors in this PR https://github.com/sympy/sympy/pull/24896 that hadn't occurred before either. We didn't change anything in the second PR related to the failing tests, so it made me think the errors are due to the new symengine release.

**Comment by moorepants** [comment]
I actually can't tell if what we saw was due to a change in symengine or changes on our end and I don't plan to investigate it further. So no worries.
